File: MarkdownIOS/SimpleMarkdownParser/Classes/Conversion/SimpleMarkdownConverter.swift
//
//  SimpleMarkdownConverter.swift
//  SimpleMarkdownParser Pod
//
//  Conversion library: utility class to convert markdown to HTML or attributed string
//

public let NSClickableTextAttributeName = "NSClickableTextAttributeName"
public let NSHighlightColorAttributeName = "NSHighlightColorAttributeName"

// Convert markdown text to an attributed string or HTML
public class SimpleMarkdownConverter {
    
    // --
    // MARK: HTML conversion handling
    // --
    
    public static func toHtmlString(fromMarkdownText: String) -> String {
        // Find symbols
        let symbolFinder = obtainSymbolFinder(forMarkdownText: fromMarkdownText)
        symbolFinder.scanText(fromMarkdownText)
        
        // Find tags from symbols and process text
        let tags = SimpleMarkdownTagFinder().findTags(text: fromMarkdownText, symbols: symbolFinder.symbolStorage.symbols)
        let processor = SimpleMarkdownTextProcessor.process(text: fromMarkdownText, tags: tags)
        
        // Process HTML
        let htmlProcessor = SimpleMarkdownHtmlProcessor.process(text: processor.text, tags: processor.tags)
        return htmlProcessor.text
    }

    
    // --
    // MARK: Attributed string conversion handling
    // --
    
    public static func toAttributedString(defaultFont: UIFont, markdownText: String, attributedStringGenerator: MarkdownAttributedStringGenerator? = nil) -> NSAttributedString {
        // Find symbols
        let symbolFinder = obtainSymbolFinder(forMarkdownText: markdownText)
        symbolFinder.scanText(markdownText)
        
        // Find tags from symbols and process text
        let stringGenerator = attributedStringGenerator ?? DefaultMarkdownAttributedStringGenerator()
        let tags = SimpleMarkdownTagFinder().findTags(text: markdownText, symbols: symbolFinder.symbolStorage.symbols)
        let processor = SimpleMarkdownTextProcessor.process(text: markdownText, tags: tags, attributedStringGenerator: stringGenerator)
        processor.rearrangeNestedTextStyles()
        
        // Set up attributed string
        let attributedString = NSMutableAttributedString(string: processor.text)
        attributedString.addAttribute(NSAttributedString.Key.font, value: defaultFont, range: NSMakeRange(0, attributedString.length))
        for index in processor.tags.indices {
            let tag = processor.tags[index]
            // Handle section spacer
            if tag.type == .sectionSpacer {
                var previousSectionTagType: MarkdownTagType?
                var nextSectionTagType: MarkdownTagType?
                var previousSectionWeight = 0
                var nextSectionWeight = 0
                for checkIndex in processor.tags.indices {
                    let checkTag = processor.tags[checkIndex]
                    if checkTag.type.isSection() {
                        if checkIndex < index {
                            previousSectionTagType = checkTag.type
                            previousSectionWeight = checkTag.weight
                        } else if checkIndex > index {
                            nextSectionTagType = checkTag.type
                            nextSectionWeight = checkTag.weight
                            break
                        }
                    }
                }
                if let previousType = previousSectionTagType, let nextType = nextSectionTagType {
                    stringGenerator.applySectionSpacerAttribute(defaultFont: defaultFont, attributedString: attributedString, previousSectionType: previousType, previousSectionWeight: previousSectionWeight, nextSectionType: nextType, nextSectionWeight: nextSectionWeight, start: tag.startPosition, length: tag.endPosition - tag.startPosition)
                }
            }
            
            // Apply attributes from tag
            stringGenerator.applyAttribute(defaultFont: defaultFont, attributedString: attributedString, type: tag.type, weight: tag.weight, start: tag.startPosition, length: tag.endPosition - tag.startPosition, extra: tag.link ?? "")
        }
        return attributedString
    }

    
    // --
    // MARK: Obtain parser instance
    // --
    
    private static func obtainSymbolFinder(forMarkdownText: String) -> SimpleMarkdownSymbolFinder {
        return SimpleMarkdownSymbolFinderSwift()
    }

}

File: MarkdownAndroid/SimpleMarkdownParserLib/src/main/java/com/crescentflare/simplemarkdownparser/conversion/SimpleMarkdownConverter.java
package com.crescentflare.simplemarkdownparser.conversion;

import android.text.SpannableStringBuilder;
import android.text.Spanned;

import com.crescentflare.simplemarkdownparser.symbolfinder.SimpleMarkdownSymbolFinder;
import com.crescentflare.simplemarkdownparser.symbolfinder.SimpleMarkdownSymbolFinderJava;
import com.crescentflare.simplemarkdownparser.symbolfinder.SimpleMarkdownSymbolFinderNative;
import com.crescentflare.simplemarkdownparser.tagfinder.MarkdownTag;
import com.crescentflare.simplemarkdownparser.tagfinder.ProcessedMarkdownTag;
import com.crescentflare.simplemarkdownparser.tagfinder.SimpleMarkdownTagFinder;

import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * Simple markdown parser library: markdown text converter
 * Convert markdown to other formats usable for Android (like html or spannable strings)
 */
public class SimpleMarkdownConverter {

    // --
    // Static member to determine availability of the native core symbol finder implementation
    // --

    private static int nativeParserLibraryLoaded = 0;


    // --
    // HTML conversion handling
    // --

    @NotNull public static String toHtmlString(@NotNull String markdownText) {
        // Find symbols
        SimpleMarkdownSymbolFinder symbolFinder = obtainSymbolFinder(markdownText);
        symbolFinder.scanText(markdownText);

        // Find tags from symbols and process text
        SimpleMarkdownTagFinder tagFinder = new SimpleMarkdownTagFinder();
        List<MarkdownTag> tags = tagFinder.findTags(markdownText, symbolFinder.getSymbolStorage().symbols);
        SimpleMarkdownTextProcessor processor = SimpleMarkdownTextProcessor.process(markdownText, tags);

        // Process HTML
        SimpleMarkdownHtmlProcessor htmlProcessor = SimpleMarkdownHtmlProcessor.process(processor.text, processor.tags);
        return htmlProcessor.text;
    }


    // --
    // Spannable conversion handling
    // --

    @NotNull public static Spanned toSpannable(@NotNull String markdownText) {
        return toSpannable(markdownText, new DefaultMarkdownSpanGenerator());
    }

    @NotNull public static Spanned toSpannable(@NotNull String markdownText, @NotNull MarkdownSpanGenerator spanGenerator) {
        // Find symbols
        SimpleMarkdownSymbolFinder symbolFinder = obtainSymbolFinder(markdownText);
        symbolFinder.scanText(markdownText);

        // Find tags from symbols and process text
        SimpleMarkdownTagFinder tagFinder = new SimpleMarkdownTagFinder();
        List<MarkdownTag> tags = tagFinder.findTags(markdownText, symbolFinder.getSymbolStorage().symbols);
        SimpleMarkdownTextProcessor processor = SimpleMarkdownTextProcessor.process(markdownText, tags, spanGenerator);
        processor.rearrangeNestedTextStyles();

        // Set up spannable
        SpannableStringBuilder spannableString = new SpannableStringBuilder(processor.text);
        for (int index = 0; index < processor.tags.size(); index++) {
            // Handle section spacer
            ProcessedMarkdownTag tag = processor.tags.get(index);
            if (tag.type == MarkdownTag.Type.SectionSpacer) {
                MarkdownTag.Type previousSectionTagType = null;
                MarkdownTag.Type nextSectionTagType = null;
                int previousSectionWeight = 0;
                int nextSectionWeight = 0;
                for (int checkIndex = 0; checkIndex < processor.tags.size(); checkIndex++) {
                    ProcessedMarkdownTag checkTag = processor.tags.get(checkIndex);
                    if (checkTag.type.isSection()) {
                        if (checkIndex < index) {
                            previousSectionTagType = checkTag.type;
                            previousSectionWeight = checkTag.weight;
                        } else if (checkIndex > index) {
                            nextSectionTagType = checkTag.type;
                            nextSectionWeight = checkTag.weight;
                            break;
                        }
                    }
                }
                if (previousSectionTagType != null && nextSectionTagType != null) {
                    spanGenerator.applySectionSpacerSpan(spannableString, previousSectionTagType, previousSectionWeight, nextSectionTagType, nextSectionWeight, tag.startPosition, tag.endPosition);
                }
            }

            // Apply span from tag
            String extra = "";
            if (tag.type == MarkdownTag.Type.OrderedListItem || tag.type == MarkdownTag.Type.UnorderedListItem) {
                extra = spanGenerator.getListToken(tag.type, tag.weight, tag.counter);
            } else if (tag.link != null) {
                extra = tag.link;
            }
            spanGenerator.applySpan(spannableString, tag.type, tag.weight, tag.startPosition, tag.endPosition, extra);
        }
        return spannableString;
    }


    // --
    // Obtain symbol finder instance based on requirements
    // --

    private static SimpleMarkdownSymbolFinder obtainSymbolFinder(String text) {
        if (nativeParserLibraryLoaded == 0) {
            try {
                System.loadLibrary("simplemarkdownparser_native");
                nativeParserLibraryLoaded = 1;
            } catch (Throwable t) {
                nativeParserLibraryLoaded = -1;
            }
        }
        return text.length() > 128 && nativeParserLibraryLoaded == 1 ? new SimpleMarkdownSymbolFinderNative() : new SimpleMarkdownSymbolFinderJava();
    }
}
